src/ConsoleArgumentBag.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console;
+
+use Tempest\Console\Exceptions\UnresolvedArgumentsException;
+
+final class ConsoleArgumentBag
+{
+    /** @var ConsoleInputArgument[] */
+    protected array $arguments = [];
+
+    /** @var string[] */
+    protected array $path = [];
+
+    /**
+     * @param array<string|int, mixed> $arguments
+     */
+    public function __construct(array $arguments)
+    {
+        $this->path = array_filter([
+            $arguments[0] ?? null,
+            $arguments[1] ?? null,
+        ]);
+
+        unset($arguments[0], $arguments[1]);
+
+        foreach (array_values($arguments) as $position => $argument) {
+            if (str_starts_with($argument, '--')) {
+                $parts = explode('=', str_replace('--', '', $argument));
+
+                $key = $parts[0];
+
+                $this->set($key, $parts[1] ?? true);
+
+                continue;
+            }
+
+            $this->set($position, $argument);
+        }
+    }
+
+    /**
+     * @return ConsoleInputArgument[]
+     */
+    public function all(): array
+    {
+        return $this->arguments;
+    }
+
+    public function get(string|int $name): ?ConsoleInputArgument
+    {
+        return $this->arguments[$name] ?? null;
+    }
+
+    public function has(string $name): bool
+    {
+        return array_key_exists($name, $this->arguments);
+    }
+
+    public function set(string|int $name, mixed $value): ConsoleInputArgument
```

Comment:
Should `set` be public?

Reply:
it depends on how this class will evolve tbh - for now most of these methods could be private.

In the initial "big PR" we also had the concept of injected parameters, like -v, -f or -h.

I think it'd be an interesting concept if we had a "middleware" layer that'd be able to inject parameters or change their values.

If we are creating a CLI application that interacts with 3rd parties and requires authentication, a middleware layer could be useful for all of the commands in our app and could inject a parameter --apiKey if it's present in storage or something.


So the flow would be:

1. Create argument bag
2. Run middlewares
3. Validate that the argument count is correct
4. Run injections (-h, -v)
5. Validate the argument bag against schema from command's handler definition
6. Execute command

Lemme know what you think about that - to me it's a super neat concept.

Reply:
That _is_ a neat concept. 👍